text/0043-block-metadata.md

```diff
@@ -0,0 +1,193 @@
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Guide-level explanation](#guide-level-explanation)
+- [Reference-level explanation](#reference-level-explanation)
+  - [Protobuf schema](#protobuf-schema)
+  - [Signing metadata](#signing-metadata)
+  - [Signing key history](#signing-key-history)
+  - [Historical AVRs](#historical-avrs)
+  - [Block Version bump](#block-version-bump)
+- [Drawbacks](#drawbacks)
+- [Rationale and alternatives](#rationale-and-alternatives)
+  - [Alternatives considered](#alternatives-considered)
+    - [Write metadata for older blocks](#write-metadata-for-older-blocks)
+- [Prior art](#prior-art)
+- [Unresolved questions](#unresolved-questions)
+- [Future possibilities](#future-possibilities)
+
+# Summary
+[summary]: #summary
+
+Add metadata to the blockchain, with consensus nodes reporting the Attested
+Verification Report and quorum set at the time a block was externalized.
+
+# Motivation
+[motivation]: #motivation
+
+This is a long-awaited feature for the MobileCoin blockchain.
+
+The AVR proves that the block was signed by an enclave -- the pubkey used to
+sign the block is in the AVR. Signing the metadata with the SCP message signing
+key proves that the block and metadata were generated by the holder of that
+private key.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When consensus nodes externalize a block, they will include `BlockMetadata` with
+the AVR and quorum set for that block.
+
+LedgerDB and WatcherDB also track and persist the block metadata.
+
+Note that while the `BlockMetadata` can be used to verify chain integrity, it
+couldn't be used to falsify any new transactions or blocks.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Add a `message BlockMetadata` to `blockchain.ArchiveBlock`, and an analogous
+`struct BlockMetadata` to `mc_blockchain_types::BlockData`.
+
+## Protobuf schema
+
+```protobuf
+message BlockMetadataContents {
+    /// The Block ID.
+    BlockID id = 1;
+
+    /// Quorum set configuration at the time of externalization.
+    QuorumSet quorum_set = 2;
+
+    /// IAS report for the enclave which generated the signature.
+    VerificationReport verification_report = 3;
+}
+
+message BlockMetadata {
+    /// Metadata signed by the consensus node.
+    BlockMetadataContents contents = 1;
+
+    /// Message signing key.
+    Ed25519Public node_key = 2;
+
+    /// Signature using `node_key` over the Digestible encoding of `contents`.
+    Ed25519Signature signature = 3;
+}
+
+// NB: Only addition is field #4.
+message ArchiveBlockV1 {
+    /// Block.
+    Block block = 1;
+
+    /// Contents of the block.
+    BlockContents block_contents = 2;
+
+    /// Block signature.
+    BlockSignature signature = 3;
+
+    /// Additional signed metadata about this block.
+    BlockMetadata metadata = 4;
+}
+```
+
+Since this is a protobuf-additive change, we don't need a new `ArchiveBlock`
```

Comment:
Would it make sense to have a blockversion field here to know if we expect the new archiveblock format with the added metadata vs the old format without it?

Reply:
It's already included, under `ArchiveBlock.v1.block.version`